fl4health/clients/ditto_client.py

```diff
@@ -0,0 +1,361 @@
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.initial_global_tensors: List[torch.Tensor]
+        self.lam = lam
+        self.global_model: nn.Module
+
+    def get_optimizer(self, config: Config) -> Dict[str, Optimizer]:
+        """
+        Returns a dictionary with global and local optimizers with string keys 'global' and 'local' respectively.
+        """
+        raise NotImplementedError
+
+    def set_optimizer(self, config: Config) -> None:
+        """
+        Ditto requires an optimizer for the global model and one for the local model. This function simply ensures that
+        the optimizers setup by the user have the proper keys and that there are two optimizers.
+
+        Args:
+            config (Config): The config from the server.
+        """
+        optimizers = self.get_optimizer(config)
+        assert isinstance(optimizers, dict) and set(("global", "local")) == set(optimizers.keys())
+        self.optimizers = optimizers
+
+    def setup_client(self, config: Config) -> None:
+        """
+        Set dataloaders, optimizers, parameter exchangers and other attributes derived from these.
+        Then set initialized attribute to True.
+
+        Args:
+            config (Config): The config from the server.
+        """
+        # Need to setup the global model here as well. It should be the same architecture as the local model so
+        # we reuse the get_model call. We explicitly send the model to the desired device. This is idempotent.
+        self.global_model = self.get_model(config).to(self.device)
+        # The rest of the setup is the same
+        super().setup_client(config)
+
+    def get_ditto_drift_loss(self) -> torch.Tensor:
+        """
+        Compute the L2 inner product between the initial global weights for the round and the current local model
+            weights. This loss function is added to the loss function for the local model when back propagating.
+
+        Returns:
+            torch.Tensor: Returns the L2 inner product between the initial global weights of the round and the current
+                local model weights.
+        """
+        assert self.initial_global_tensors is not None and self.lam is not None
+        # Using parameters to ensure the same ordering as exchange
+        local_model_weights = [layer_weights for layer_weights in self.model.parameters()]
+        assert len(self.initial_global_tensors) == len(local_model_weights)
+        assert len(self.initial_global_tensors) > 0
+
+        layer_inner_products: List[torch.Tensor] = [
+            torch.pow(torch.linalg.norm(initial_layer_weights - iteration_layer_weights), 2.0)
+            for initial_layer_weights, iteration_layer_weights in zip(self.initial_global_tensors, local_model_weights)
+        ]
+
+        # network l2 inner product tensor weighted by lambda
+        return self.lam * torch.stack(layer_inner_products).sum()
```

Comment:
Should this be self.lam/2 to be consistent with paper (and [implementation](https://github.com/litian96/ditto/blob/26d9b29a62f3872dcdf071be08e897607c283bb4/flearn/trainers_MTL/ditto.py#L117))? Without it we would be scaling the contribution of the regularization by 2 which isn't a huge deal but if we are using the paper to get the space of lam to iterate over as hyperparam we should scale the loss or lams accordingly.

Reply:
100%, that was an oversight by me. Will update this and the tests accordingly.

Reply:
I can do it if you like! Don't want to make you do it on vacation haha! Just let me know

Reply:
Already in progress 🙂

Reply:
Other than that, everything looks good to me! So I will take a look at the changes quickly and we can merge it in to get the ball rolling for experiments

Reply:
Sounds good. Make sure you take a little bit of time to verify the FLamby scripts in the other PR. I've gone over them a few times, but it helps to have a fresh set of eyes, just in case I've gone blind to anything.

Reply:
Yeah I will definitely spend some time on the next PR as well! Any feedback should come in the next few hours :)